### ISSUE TITLE
3 ms delays in some ABT_thread_create() calls

### ISSUE THREAD

**Issue by carns** [pmodels/argobots#369]
We have a Mochi test case (contributed by @philip-davis) with some unexpected stalls in it.  I'll link the discussion from the Mochi mailing list below, but we've narrowed down the root cause in more detail as described below.

This is the configuration from an Argobots perspective:
- there is 1 "producer" ES with it's own pool
- there are 4 "consumer" ESs in a different pool
- the producer periodically calls `ABT_thread_create()` to spawn detached threads that will be executed by the consumers.
- `ABT_MAX_NUM_STACKS` is set to 8
- `ABT_THREAD_STACKSIZE` is set to 2 MiB
- I'm running the test case on the ALCF Cooley system, but @philip-davis originally triggered it elsewhere

(For some background: I know that these config parameters are unusual for Argobots, but they are intentional in the context of this use case. The stack size is very large to accommodate invoking deep userspace network library functions, and the num stacks is small to constrain memory usage in these kinds of producer/consumer cases since the threads are retired from a different ES from where they are issued.  These are conservative values that could be adjusted, though.)

At any rate, what we observe is that the `ABT_thread_create()` calls will occasionally consume roughly 3 milliseconds of time *before* actually inserting the new thread into the consumer pool.  Once the thread hits the pool, everything then executes in a timely manner.

The memory management within Argobots is a little complex to trace quickly, but the unique thing about the slow `ABT_thread_create()` invocations is that they trigger calls to `ABTU_alloc_largepage()` within Argobots, which on this system eventually translates to an `mmap()` call.  The allocation size 8 MiB, and it is either invoked once or twice within the `ABT_thread_create()` call path.  I would assume that these happen at points in time when the pool of preallocated stacks is exhausted and more memory has to be allocated.

Obviously we could tune some parameters to adjust memory usage, but fundamentally even allocating 8 MiB of memory with an anonymous mmap shouldn't take 3 ms, so I think there is some more fundamental performance bug in this code path.

My suspicion (I have no evidence; this is just a hunch based on the magnitude of the delay) is that Argobots is holding a spinlock in this allocation path.  the `mmap()` is triggering a system call that context switches while the spinlock is held, potentially causing it to get stuck until the kernel scheduler time slice expires.

I'll keep investigating to see if I can either spot the problem or else produce a more compact Argobots-specific reproducer, but I wanted to go ahead and describe what I'm seeing in case someone has a suggestion in the mean time.

**Comment by carns** [comment]
Oh, the original discussion that prompted this investigation is at https://lists.mcs.anl.gov/pipermail/mochi-devel/2021-October/000126.html.  We originally suspected a network problem but after investigating further (off list) we tracked the delays down to argobots thread creation.

**Comment by carns** [comment]
FWIW looking closer it looks like the `mmap()` calls themselves are comparatively fast (a few usec).  It seems like there is something else in the surrounding path when Argobots needs to trigger an allocation that is getting stuck.

**Comment by shintaro-iwasaki** [comment]
Thanks for reporting an issue!

> I would assume that these happen at points in time when the pool of preallocated stacks is exhausted and more memory has to be allocated.

Yes, if the underlying execution stream is running out of cached ULT stacks, it calls `ABTU_alloc_largepage()`. `ABTU_alloc_largepage()` by default uses `mmap()`.

I am not sure if this issue is because of the Argobots memory pool implementation.  It uses a lock-free implementation (a lock-free LIFO for global pool), so it should not cause an unintentionally large delay.
https://github.com/pmodels/argobots/blob/main/src/mem/mem_pool.c#L255-L269

A rough idea of this algorithm is as follows (assuming that `ABT_MEM_MAX_NUM_STACKS=8` and `ABT_THREAD_STACKSIZE=$((2 * 1024 * 1024))` are set).

```c
ABTI_mem_pool_header *p_header = NULL;
int count = 0;
const int BUCKET_SIZE = 4; // It will be 4 for the given envvars.
while (count < BUCKET_SIZE) {
    if (global_memory_pool is empty) { // lock-free
        mem = ABTU_alloc_largepage();
        [Create a linked list and connect it to p_header];
        count += mem / ABT_THREAD_STACKSIZE;
    } else {
        [Take some from global_memory_pool]; // lock-free
    }
    if (count > BUCKET_SIZE) // Took too much.  Return it.
        return_to_memory_pool(p_header, count - BUCKET_SIZE); // lock-free
}
return p_header;
```

This lock-free implementation relies on 128-bit atomic compare-and-swap, but modern Intel machines should support this (https://github.com/pmodels/argobots/blob/main/src/include/asm/abtd_asm_int128_cas.h#L16-L36).  Note that, if the underlying machine does not support 128-bit atomic CAS, the algorithm gets not lock-free because the implementation falls back to a spin-lock.

The following would also be suspicious.

1. Memory overheads
> Obviously we could tune some parameters to adjust memory usage, but fundamentally even allocating 8 MiB of memory with an anonymous mmap shouldn't take 3 ms, so I think there is some more fundamental performance bug in this code path.

I don't think `mmap()` usually takes that long, unless `mmap()` uses a global spin-lock internally and there's a lock conflict.

Note that Argobots' memory-pool algorithm touches memory to create a linked list.  With the given envvars, it should access memory four (and maybe plus a few) times.

2. Pool operation overheads

Assuming that `WAIT_POOL` is used for this workload.  If the other execution streams are sleeping on a condition variable and `ABT_thread_create()` needs to wake them up (https://github.com/pmodels/argobots/blob/main/src/pool/fifo_wait.c#L131-L134), the mutex/cond overhead can be sometimes very large depending on the underlying POSIX mutex/cond implementation.

---

Argobots can disable memory pool (`--disable-mem-pool`), which falls back to `malloc()`/`free()` per ULT stack.  This would be useful to identify an issue.  If the memory pool is the problem but  Glibc's `malloc()` is too slow, using an efficient `malloc()` implementation (e.g., jemalloc or tcmalloc) would alleviate the issue.

I know the comment above does not identify the issue. In any case, I am happy to investigate this issue. Please let me know any updates.

**Comment by carns** [comment]
Thanks for the comments/ideas @shintaro-iwasaki !

I've already ruled out mmap() itself and the `fifo_wait` pool as causes for this particular delay.  I'll look into the other possibilities though.  I've been thinking about it and I believe that I can probably make a stand-alone reproducer that will make it easier to experiment with.

**Comment by carns** [comment]
[abt-thread-create-stall.tar.gz](https://github.com/pmodels/argobots/files/7516668/abt-thread-create-stall.tar.gz)

I've attached a small reproducer.  It measures the latency of individual `ABT_thread_create()` calls with the configuration originally described in this issue, except that it uses the normal default pool and scheduler to eliminate potential influence from the `fifo_wait` pool.  The attached tar ball also includes example output from two systems (Cooley@ALCF and my own laptop), also shown below for discussion convenience.

On my laptop: The example behaves as expected.  There is an extra delay on every 4th thread create call due to the memory allocation.  Obviously I'd like for those delays to be lower, but now that I know how to observe them, I'm pretty confident that we can tune it.  I'll circle back to that later.
```
# <iteration>	<thread_create latency (usec)>
0	0.953674
1	0.238419
2	0.238419
3	21.219254
4	0.000000
5	0.476837
6	0.238419
7	12.874603
8	0.238419
9	0.476837
10	0.238419
11	8.583069
# SLOWEST ITERATION: 3 at 21.219254 usec

# System info:
# argobots@main, default spack options, 2021-11-10
# Ubuntu 21.10
# gcc 11.2
# Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz
```

On Cooley: This is where the performance bug shows up.  Every 4th iteration is slow again, as on my laptop, but this time the delay is extreme:
```
# <iteration>	<thread_create latency (usec)>
0	2.384186
1	0.476837
2	0.238419
3	21097.183228
4	0.238419
5	0.476837
6	0.476837
7	2075.910568
8	0.238419
9	0.715256
10	0.238419
11	2078.533173
# SLOWEST ITERATION: 3 at 21097.183228 usec

# System info:
# argobots@main, default spack options, 2021-11-10
# RHEL 7.9
# gcc 8.2
# Intel(R) Xeon(R) CPU E5-2620 v3 @ 2.40GHz
```
It is consistently taking at least 2000 usec every 4th iteration, and occasionally spiking even higher.  I've measured the underlying anonymous `mmap()` cost on cooley and it is at most 10s of usec despite these large `ABT_thread_create()` delays, so it seems like there is some other problem in the allocation path.  I'd like to resolve this (because it seems like a bug) before tuning any caching/allocation parameters.

I don't yet know which difference (CPU, compiler, OS, or something else entirely) makes this so extreme on Cooley.  @philip-davis has observed very similar behavior on the Frontera system at TACC as well.

**Comment by carns** [comment]
It looks like neither Cooley nor my laptop are enabling 128 CAS operations, but I think this is actually due to a bug in the autoconf test; I'll open a separate issue to track that.

**Comment by carns** [comment]
Just a quick update here.  @philip-davis and I have been digging deeper into this and trying to isolate the delay to more specific lines of code.  We haven't found any specific problems with Argobots.  It looks more like a fundamental system memory allocation problem (that's either paid at allocation time or on subsequent memory accesses, depending on which allocator you use).  We are trying to isolate that outside of Argobots and figure out what's causing it.  We'll share what we find.

**Comment by carns** [comment]
Oh, for convenience here is the output from the reproducer on Cooley:
```
$ ./allocation-latency -m mmap -s 8388608 -S 1048576 -i 64
allocation and access times per 8388608 byte buffer:
    mean: 1872.684807 usec
    max: 2150.058746 usec
```
... and the same test case on my laptop:
```
./allocation-latency -m mmap -s 8388608 -S 1048576 -i 64
allocation and access times per 8388608 byte buffer:
    mean: 42.367727 usec
    max: 76.055527 usec
```

**Comment by shintaro-iwasaki** [comment]
Thank you for the update! Unfortunately now I don't have an access to RHEL/CentOS 7.x, but if a different access pattern gives better performance on some systems, Argobots should provide a different memory allocation mechanism. We'd appreciate it if you could post any update (but please feel free to close this issue if it'd be irrelevant to Argobots).

**Comment by carns** [comment]
I'll follow up a bit for people who might encounter this issue later:

@philip-davis uncovered that the key distinction on the systems exhibiting this problem is that they use transparent huge pages (THP) automatically.  You can check this setting as follows:

```
$ cat /sys/kernel/mm/transparent_hugepage/enabled
[always] madvise never
```
Systems such as Cooley and Frontera (RHEL 7 based, though it is not confined to that distribution) set it to `always`, which means that the kernel will automatically attempt to use huge pages for memory allocations meeting certain heuristics.  `madvise` enables THP, but only as an opt-in feature via `madvise()` flags.  `never` disables the feature.

At any rate, we can completely eliminate the abnormal delays in our reproducers by using `madvise(buffer, size, MADV_NOHUGEPAGE)` on individual allocations, or globally in the process by using `prctl(PR_SET_THP_DISABLE, 1, 0, 0, 0);`.

FWIW, it seems like the trend in newer distributions is to use the `madvise` rather than the `always` setting.  That allows applications to opt-in if they know hugepages will be helpful.  

In summary, this clearly isn't an Argobots flaw.  It just happened to get caught in the crossfire for an edge case that performs extraordinarily poorly for automatic THP systems, at least when used with Margo's preferred thread stack size.

For our specific use case we are in a good position to manage our own allocations, we are going to pursue the mechanisms discussed on the Argobots mailing list at https://lists.argobots.org/pipermail/discuss/2021-November/000162.html .
